### PR TITLE
[WIP] Counter with a random numbers

### PR DESCRIPTION
**Pull request by SiarheiNikitsinV**

### CODE REVIEW

File: src/containers/App/App.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    counter={props.counter}    
+    onAdd={props.onAdd} 
+    onSub={props.onSub}      
+    onAddRandomPositiveNumber={props.onAddRandomPositiveNumber}
+    onAddNegativeRandomNumber={props.onAddRandomNegativeNumber}
+  />
+);
+
+
+function mapStateToProps(state) {
```
Comment (mastapchyk) [inline src/containers/App/App.js:22]: create a container of the counter component, move mapsStateToProps and mapDispatchToProps to the container

File: src/components/Counter/Counter.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+  return (
+    <div className="container counterWrapper text-center">
+      <div className="counter">Counter: {counter}</div>
+      
+      <div className="counterButtons">
+        <button onClick={onAdd}>Add 1</button>
+        <button onClick={onSub}>Sub 1</button>     
+            
+        <button onClick={onAddRandomPositiveNumber}>Add random positive number</button>
```
Comment (mastapchyk) [inline src/components/Counter/Counter.js:16]: replace to addRandomAsyncNumber with argument isNegative

File: src/reducers/counterReducer.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+
+const initialState = {
+  counter: 0
+}
+  
+const counterReducer = (state = initialState, action) => {
+  switch (action.type) {    
+    case constants.ADD:
+      return {
```
Comment (mastapchyk) [inline src/reducers/counterReducer.js:11]: replace "return {...}" to lodash/fp assoc

File: src/sagas/counterSaga.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import axios from 'axios';
+
+import { put, takeEvery, all, call } from 'redux-saga/effects';
+
+const API = 'http://localhost:5000/api/number';
```
Comment (mastapchyk) [inline src/sagas/counterSaga.js:5]: replace url to proxy in webpack
google "how to use proxy in create-react-app"

File: src/components/Counter/Counter.js (hunk 2 above)
Comment (SiarheiNikitsinV) [reply src/components/Counter/Counter.js:16]: Completed

File: src/sagas/counterSaga.js (hunk 4 above)
Comment (SiarheiNikitsinV) [reply src/sagas/counterSaga.js:5]: Completed

File: src/reducers/counterReducer.js (hunk 3 above)
Comment (SiarheiNikitsinV) [reply src/reducers/counterReducer.js:11]: Completed

File: src/containers/App/App.js (hunk 1 above)
Comment (SiarheiNikitsinV) [reply src/containers/App/App.js]: Completed

File: src/containers/App/App.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import React from 'react';
+
+import Counter from '../Counter';
+
+
```
Comment (mastapchyk) [inline src/containers/App/App.js:5]: app is container?

File: src/containers/Counter/index.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import { connect } from 'react-redux';
+
+import { add, addRandomAsyncNumber } from '../../actions/counterActions';
+
+import CounterComponent from '../../components/Counter/Counter';
+
+import { getCounter } from '../../selectors/counterSelectors';
+
+
+const Counter = (props) => (
```
Comment (mastapchyk) [inline src/containers/Counter/index.js:11]: the Counter doesn't need

File: src/containers/Counter/index.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+function mapStateToProps(state) {
+  return {
+    counter: getCounter(state)    
+  }
+}
+
+function mapDispatchToProps(dispatch) {
+  return {    
+    onAdd: () => dispatch(add(1)),
```
Comment (mastapchyk) [inline src/containers/Counter/index.js:30]: rewrite without dispatch

File: src/containers/Counter/index.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    onSub: () => dispatch(add(-1)),    
+    onAddRandomPositiveNumber: () => dispatch(addRandomAsyncNumber('isPositive')),
```
Comment (mastapchyk) [inline src/containers/Counter/index.js:32]: why string args?
why don't use isNegative? default is false, a optional is true

File: src/reducers/counterReducer.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+const initialState = {
+  counter: 0
+}
+  
+const counterReducer = (state = initialState, { type, payload }) => {
+  switch (type) {    
+    case constants.ADD:      
+      return compose(
+        assoc(['counter'], state.counter + payload)
```
Comment (mastapchyk) [inline src/reducers/counterReducer.js:14]: why use compose?

File: src/selectors/counterSelectors.js [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import { createSelector } from 'reselect';
+
+
+export const getCounter = (state) => {
+  return state.counter;
+}
+
+export const getCounterSuperSelector = createSelector(getCounter, (counter) => {
```
Comment (mastapchyk) [inline src/selectors/counterSelectors.js:8]: what is super selector?
what is difference between getCounter and getCounterSuper..?

File: src/actions/counterActions.js [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import { createAction } from 'redux-actions';
+
+import * as constants from '../utils/constants/counterConstants';
+
+export const add = createAction(constants.ADD, payload => payload);
```
Comment (mastapchyk) [inline src/actions/counterActions.js:5]: why payload => payload?

File: src/actions/counterActions.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
+import { createAction } from 'redux-actions';
+
+import * as constants from '../utils/constants/counterConstants';
+
+export const add = createAction(constants.ADD, payload => payload);
+export const addRandomAsyncNumber = createAction(constants.ADD_RANDOM_ASYNC_NUMBER, payload => payload);
```
Comment (mastapchyk) [inline src/actions/counterActions.js:6]: use routine for async actions

File: src/containers/App/App.js (hunk 5 above)
Comment (SiarheiNikitsinV) [reply src/containers/App/App.js:5]: Completed

File: src/containers/Counter/index.js (hunk 8 above)
Comment (SiarheiNikitsinV) [reply src/containers/Counter/index.js:32]: Completed

File: src/actions/counterActions.js (hunk 12 above)
Comment (SiarheiNikitsinV) [reply src/actions/counterActions.js:6]: Completed

File: src/actions/counterActions.js (hunk 11 above)
Comment (SiarheiNikitsinV) [reply src/actions/counterActions.js:5]: I thought that according to the specs i need this if I want to use payloads. Just extra code in this situation. Removed. Completed

File: src/reducers/counterReducer.js (hunk 9 above)
Comment (SiarheiNikitsinV) [reply src/reducers/counterReducer.js:14]: I've just watched the TTS code and tried to do the working solution. Extra code in this situation. Removed. Completed

File: src/containers/Counter/index.js (hunk 6 above)
Comment (SiarheiNikitsinV) [reply src/containers/Counter/index.js:11]: Completed

File: src/selectors/counterSelectors.js (hunk 10 above)
Comment (SiarheiNikitsinV) [reply src/selectors/counterSelectors.js:8]: getCounter is a selector. And getCounterSuperSelector is a memoized selector made with library reselect which uses a getCounter selector as a base. By a mistake I didn't use a reselect selector. Code improved. Completed

File: src/containers/Counter/index.js (hunk 7 above)
Comment (SiarheiNikitsinV) [reply src/containers/Counter/index.js]: Completed

**Pull request merged**